### ISSUE TITLE
Quantization and training=True

### ISSUE THREAD

**Issue by Burton2000** [alibaba/TinyNeuralNetwork#159]
Hi again,

I'm currently experimenting with Quantization and see that the PostQuantizer puts models into training mode before tracing the graph. For some models I've experimented with this can cause the quantize() method to fail depending on what the model does in its forward step.

One example, DeepLabv3 model in torchvision. It contains a BatchNorm layer that works on a 256x1x1 input. The layer will throw an error if you pass a dummy input with a batch size of 1, as collecting BatchNorm stats (as we are now in training mode) for this batch size is not valid.

Other times a model might expect additional inputs to the model in the training phase such as target tensors that need to be passed along.

I see various ways around these issues and I presume the best is to just rewrite the forward method of the model to avoid problems entirely when switching between training=True and False. I wondered if you had come across any cases like this and if you have any recommendations of what you did to overcome these problems or my thinking of rewriting the forward method is the best thing to do or maybe I am doing something incorrectly?

Many thanks,
Richard

Deeplabv3 example:
```
with model_tracer():
    weights = torchvision.models.segmentation.DeepLabV3_MobileNet_V3_Large_Weights.DEFAULT
    model = torchvision.models.segmentation.deeplabv3_mobilenet_v3_large(weights=weights)
    model.eval()

    dummy_input = torch.rand((1, 3, 520, 520))  # If this is changed to (2,3,520,520) if will work

    quantizer = PostQuantizer(model, dummy_input, work_dir='out', config={'asymmetric': True, 'per_tensor': False})
    ptq_model = quantizer.quantize()
```

**Comment by peterjc123** [comment]
> I'm currently experimenting with Quantization and see that the PostQuantizer puts models into training mode before tracing the graph. For some models I've experimented with this can cause the quantize() method to fail depending on what the model does in its forward step.

You mean the `QATQuantizer`, right? The `PostQuantizer` won't put the model in training mode. For `QATQuantizer`, we need to ensure that the model is in training mode, otherwise it won't work when calling `torch.quantization.fuse_modules`. See 
https://github.com/pytorch/pytorch/blob/v1.10.0/torch/ao/quantization/fuser_method_mappings.py#L59.


> I see various ways around these issues and I presume the best is to just rewrite the forward method of the model to avoid problems entirely when switching between training=True and False. I wondered if you had come across any cases like this and if you have any recommendations of what you did to overcome these problems or my thinking of rewriting the forward method is the best thing to do or maybe I am doing something incorrectly?

Currently, the best way is to rewrite the forward by oneself. We plan to work on a better way that overwrites the forward function on the fly, so that we can generate code for both training and evaluation.

> Deeplabv3 example:
> 
> ```
> with model_tracer():
>     weights = torchvision.models.segmentation.DeepLabV3_MobileNet_V3_Large_Weights.DEFAULT
>     model = torchvision.models.segmentation.deeplabv3_mobilenet_v3_large(weights=weights)
>     model.eval()
> 
>     dummy_input = torch.rand((1, 3, 520, 520))  # If this is changed to (2,3,520,520) if will work
> 
>     quantizer = QATQuantizer(model, dummy_input, work_dir='out', config={'asymmetric': True, 'per_tensor': False})
>     ptq_model = quantizer.quantize()
> ```

Actually, this is okay to pass a tensor with `batch_size > 1` since dynamic shape are well handled during Graph Tracing and Quantization Rewrite. You are free to pass tensors with arbitrary batch size during QAT afterwards.

**Comment by Burton2000** [comment]
> You mean the QATQuantizer, right? The PostQuantizer won't put the model in training mode. 

Hmm but the PostQuantizer doesn't have its own quantize method, it instead calls the [base method](https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/graph/quantization/quantizer.py#L318) quantize from QATQuantizer, which as you say puts the model into training mode as needed for QAT. Or did I miss something again?

> Currently, the best way is to rewrite the forward by oneself. We plan to work on a better way that overwrites the forward function on the fly, so that we can generate code for both training and evaluation. But for your case (given in the code example), I didn't see why you need to do that, as the code logic for training and evaluation is just the same.

Thanks that's good to know!

> Actually, this is okay to pass a tensor with batch_size > 1 since dynamic shape are well handled during Graph Tracing and Quantization Rewrite. You are free to pass tensors with arbitrary batch size during QAT afterwards.

Ah yes you are correct I should have spotted this! The solution was using batch_size > 1 for the quantization dummy input and use a different dummy input in conversion with batch_size=1 and everything works as I wanted.

Thanks again for your time and help its really appreciated!

**Comment by peterjc123** [comment]
> Hmm but the PostQuantizer doesn't have its own quantize method, it instead calls the [base method](https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/graph/quantization/quantizer.py#L318) quantize from QATQuantizer, which as you say puts the model into training mode as needed for QAT. Or did I miss something again?

Double checked, and yeah, it's a bug. We will put up a fix soon.

**Comment by peterjc123** [comment]
> > Hmm but the PostQuantizer doesn't have its own quantize method, it instead calls the [base method](https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/graph/quantization/quantizer.py#L318) quantize from QATQuantizer, which as you say puts the model into training mode as needed for QAT. Or did I miss something again?
> 
> Double checked, and yeah, it's a bug. We will put up a fix soon.

Should be fixed via 4c2536f3f5bc96485efb92cb6aed54fae77acabc

**Comment by Burton2000** [comment]
Just tried it and an issue I had with YoloX is now fixed as a result thank you!
